ttnn/cpp/ttnn/operations/experimental/cnn/convert_to_hwc/device/convert_to_hwc_program_factory.cpp

```diff
@@ -0,0 +1,223 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "convert_to_hwc_program_factory.hpp"
+
+#include "tt-metalium/tt_backend_api_types.hpp"
+
+#include "ttnn/tensor/types.hpp"
+#include "ttnn/operations/data_movement/sharded/sharded_common.hpp"
+
+namespace ttnn::operations::experimental::cnn::detail {
+
+using namespace tt::constants;
+
+tt::tt_metal::operation::ProgramWithCallbacks multi_core_convert_to_hwc(const Tensor& a, Tensor& output) {
+    tt::tt_metal::Program program = tt::tt_metal::CreateProgram();
+
+    // If we are pulling from DRAM we infer the shard shape by using the output shard spec
+    const bool is_input_in_dram = a.buffer()->core_type() == CoreType::DRAM;
+    const auto input_shard_height = is_input_in_dram ? output.shard_spec()->shape[1] : a.shard_spec()->shape[0];
+    const auto input_shard_width = is_input_in_dram ? output.shard_spec()->shape[0] : a.shard_spec()->shape[1];
+    const auto input_core_grid = is_input_in_dram ? output.shard_spec()->grid : a.shard_spec()->grid;
+    const auto input_cores = corerange_to_cores(
+        input_core_grid, std::nullopt, a.shard_spec()->orientation == tt::tt_metal::ShardOrientation::ROW_MAJOR);
+
+    TT_FATAL(input_shard_height <= TILE_HEIGHT, "Shard height must be 32 or smaller");
+    TT_FATAL(input_shard_width % TILE_WIDTH == 0, "Shard width must be multiple of tile width");
+
+    const auto create_circular_buffer = [&program, &input_core_grid](
```

Comment:
Consider using existing utility 
tt::tt_metal::create_cb 
instead of rolling your own